### PR TITLE
Add Sprite movement

### PR DESCRIPTION
**Pull request by robegamesios**
hi @jnthas i have a working prototype of allowing the sprite movement. 

here is the sample json. in this clockface, the cube bumps up every 10 seconds same time as mario jumps. could use some more improvements. 

[mario-1-moving.json](https://github.com/jnthas/cw-cf-0x07/files/13172998/mario-1-moving.json)

https://github.com/jnthas/cw-cf-0x07/assets/10041871/61594be4-078f-4639-8a21-48a7d48b27f5

### CODE REVIEW

File: Clockface.cpp [hunk 1]
```diff
@@ -157,34 +152,104 @@ void Clockface::createSprites()
...
-      //     (s.pace < 0 && s.x <= value["x1"].as<const int8_t>() && s.y <= value["y1"].as<const int8_t>()))) {
-
-      //   s.x = value["x"].as<const int8_t>();
-      //   s.y = value["y"].as<const int8_t>();
-      // }
+  for (auto& sprite : sprites) {
+    uint8_t totalFrames = sprite->_totalFrames;
+    uint32_t loopDelay = doc["loop"][sprite->_spriteReference]["loopDelay"].as<uint32_t>() ?: delay;
+    uint16_t frameDelay = doc["loop"][sprite->_spriteReference]["frameDelay"].as<uint16_t>() ?: delay;
+    unsigned long moveStartTime = doc["loop"][sprite->_spriteReference]["moveStartTime"].as<unsigned long>() ?: 1;
```
Comment (robegamesios) [inline Clockface.cpp:163]: fyi, PR includes my first PR since i merged my updates to my main.

File: Clockface.cpp [hunk 2]
```diff
@@ -157,34 +152,104 @@ void Clockface::createSprites()
...
+
+          } else {
+            // Movement is complete
+            sprite->setX(sprite->getMoveTargetX());
+            sprite->setY(sprite->getMoveTargetY());
+            renderImage(doc["sprites"][sprite->_spriteReference][sprite->_currentFrame]["image"].as<const char *>(), sprite->getX(), sprite->getY());
+              
+            // Check if the sprite should return to its initial position
+            if (sprite->shouldReturnToOrigin()) {
+              sprite->startMoving(sprite->getMoveInitialX(), sprite->getMoveInitialY(), sprite->getMoveDuration(), false);
```
Comment (robegamesios) [inline Clockface.cpp:217]: this causes the bug, for some reason the code calls this multiple times. e.g. progress is more than 1.

**Comment by jnthas** [comment]
Sorry the delay, Rob. I'm going to merge the changes and test it soon.

**Comment by robegamesios** [comment]
> Sorry the delay, Rob. I'm going to merge the changes and test it soon.

no worries. i have another branch wherein i packed in the clock themes into one, plus an audio visualizer. the problem is i get a no memory error if the audio visualizer and canvas theme are in the same code. there is something in the canvas theme that's not happy, maybe you can help me figure it out? here are  the branches, thanks: 

https://github.com/robegamesios/clockwise/tree/feature/add_audio_visualizer
https://github.com/robegamesios/clockwise/tree/feature/add_canvas

**Comment by jnthas** [comment]
Got it. You added all themes in a single firmware. It's too audacious :D, ESP32 memory just can't. And it doesn't scale. To have all together I see two options:
- Via hardware, increase memory via flash (that winbond chips) or sdcard, so all resources we load with PROGMEM like images could be allocated there
- Via software: improve the Canvas engine so that it could render almost everything via a resource file

**Comment by robegamesios** [comment]
> Got it. You added all themes in a single firmware. It's too audacious :D, ESP32 memory just can't. And it doesn't scale. To have all together I see two options:
> 
> * Via hardware, increase memory via flash (that winbond chips) or sdcard, so all resources we load with PROGMEM like images could be allocated there
> * Via software: improve the Canvas engine so that it could render almost everything via a resource file

yeah it will not scale for sure and packing themes will require new bin files, im just doing it as an experiment to figure out the fastest way to change themes.

when i compared the bin file sizes, having a single theme is about 900kb, whilst packing them all is about 1.2mb which is 94% of the allowed memory. the esp32 still handles it if I want to batch the themes as separate bin files. then you can change the theme set via OTA.

between the hardware and software update, i vote for the software update. the Canvas engine has potential, just need to work out the memory issues im encountering. which part of the canvas engine allocates a lot of memory? i know the downloading part eats up a lot of memory due to the dynamicJsonDocument.  i tried using a different deserializer that streams but i ended up with the same problem, i ran out of memory, since i need to assign variables to hold the data. the base64 strings in the json are long and i don't think there is a way to shorten those. thanks

**Comment by jnthas** [comment]
I liked the idea to have compiled themes as BIN files, it solves the issue with Canvas you asked. Probably JSON handling is the most expensive in terms of memory, it consumes a lot not only flash but also RAM. OTA is a desired feature for other users, it's in the roadmap to develop something but it needs to be planned.

**Comment by robegamesios** [comment]
those 2 branches i mentioned above implements OTA, if you load the audio visualizer bin, when you click on the canvas theme in the web dashboard, it will OTA the canvas firmware + all the other themes except audio visualizer. it also shows a progress of the OTA in the led matrix. It goes the same if you loaded the canvas firmware then clicked on the audio visualizer in the web dashboard, the only problem here is there is no progress indicator in the matrix panel. for some reason, i keep getting error of not being able to upload if i try to show a progress indicator. there is something in the canvas engine that i still dont understand that is causing these panic issues.

**Comment by jnthas** [comment]
Might it be because of WiFiClientSecure? It's only used on Canvas to access the JSON file on GitHub

**Pull request merged**